pkg/apis/extensions/types.go

```diff
@@ -809,3 +810,80 @@ type PodSecurityPolicyList struct {
 
 	Items []PodSecurityPolicy `json:"items"`
 }
+
+// +genclient=true
+
+// Template defines a list of partially complete Objects and a list of Parameters that
+// can be processed into a Config by substituting parameterized values at processing time.
+// Parameter values maybe set at processing time.
+type Template struct {
+	unversioned.TypeMeta `json:",inline"`
+	api.ObjectMeta       `json:"metadata,omitempty"`
+
+	Spec TemplateSpec `json:"spec"`
+}
+
+type TemplateSpec struct {
+	// Parameters is a list of Parameters used to substitute values into Template Objects
+	Parameters []Parameter `json:"parameters,omitempty"`
+
+	// Objects is a list of partially complete Objects with substitution symbols.
+	Objects []runtime.RawExtension `json:"objects"`
+}
+
+// TemplateList is a list of Template objects.
+type TemplateList struct {
+	unversioned.TypeMeta `json:",inline"`
+	unversioned.ListMeta `json:"metadata,omitempty"`
+	Items                []Template `json:"items"`
+}
+
+const (
+	StringParam = "string"
+	IntParam    = "integer"
+	BoolParam   = "boolean"
+	Base64Param = "base64"
+)
+
+// Parameter defines a name/value variable that is substituted into Template Objects at
+// Template processing time.
+type Parameter struct {
+	// Name defines the symbol to be replaced.  Name should appear as $(Name) in Objects.
+	Name string `json:"name"`
+
+	// DisplayName is used instead of Name when displaying the Parameter in a UI
+	DisplayName string `json:"displayName,omitempty"`
+
+	// Description is used to give context about the Parameter, such as its purpose
+	// and what values are acceptable.
+	Description string `json:"description,omitempty"`
+
+	// Value holds a default value to replace all occurrences of $(Name) within the Template's
+	// Objects when the Template is processed.  The Value maybe overridden at Template
+	// processing time.  If no Value is provided as a default or override, then
+	// the empty string will be used for substitution.
+	Value string `json:"value,omitempty"`
+
+	// Required indicates the parameter must have a non-empty value when the Template is processed.
+	// Parameters of Type 'integer' and 'boolean' are always Required.
+	Required bool `json:"required,omitempty"`
+
+	// Type is the type that the parameter value must be parsed to. Type may be one of
+	// 'string', 'integer', 'boolean', or 'base64'.
+	// Type is used by clients to provide validation of user input and direction to users.
+	//  Parameters used to define integer or boolean fields (e.g. replicaCount) should have the
+	// Type set to integer or boolean accordingly.
+	Type string `json:"type,omitempty"`
+}
+
+// TemplateParameters contains the substitution parameter overrides when processing a Template
+type TemplateParameters struct {
+	unversioned.TypeMeta `json:",inline"`
+
+	// Name is the name of the Template to be processed.
+	Name string `json:"name"`
```

Comment:
I don't want any fields here that aren't actually needed/supported.

In retrospect, ObjectMeta was confusing, since most of the fields aren't usable, and it refers to the parent resource rather than the subresource. A subset of ObjectReference probably would have been more appropriate. 

I'm open to anything that would make programmatic clients easier to write (e.g., used with the discovery API), but I don't envision users writing declarative config files for subresources.